source/multigrid/mg_transfer_prebuilt.cc

```diff
@@ -326,40 +346,131 @@ void MGTransferPrebuilt<VECTOR>::build_matrices (
+                {
+                  copy_indices[level].push_back(
+                    std::make_pair (global_dof_indices[i], level_dof_indices[i]));
+                }
               else if (global_mine)
-                copy_indices_to_me[level].push_back(
-                  std::make_pair (global_dof_indices[i], level_dof_indices[i]));
+                {
+                  copy_indices_global_mine[level].push_back(
+                    std::make_pair (global_dof_indices[i], level_dof_indices[i]));
+
+                  //send this to the owner of the level_dof:
+                  send_data_temp.push_back(dof_pair(level, global_dof_indices[i], level_dof_indices[i]));
+                }
               else
-                continue;
+                {
+                  // somebody will send those to me
+                }
 
               dof_touched[global_idx] = true;
             }
         }
     }
 
+  const dealii::parallel::distributed::Triangulation<dim,spacedim> *tria =
+    (dynamic_cast<const parallel::distributed::Triangulation<dim,spacedim>*>
+     (&mg_dof.get_tria()));
+  AssertThrow(send_data_temp.size()==0 || tria!=NULL, ExcMessage("parallel Multigrid only works with a distributed Triangulation!"));
+
+  if (tria)
+    {
+      // TODO: Searching the owner for every single DoF becomes quite
+      // inefficient. Please fix this, Timo.
+      std::set<unsigned int> neighbors = tria->level_ghost_owners();
+      std::map<int, std::vector<dof_pair> > send_data;
+
+      // * find owners of the level dofs and insert into send_data accordingly
+      for (typename std::vector<dof_pair>::iterator dofpair=send_data_temp.begin(); dofpair != send_data_temp.end(); ++dofpair)
+        {
+          for (std::set<unsigned int>::iterator it = neighbors.begin(); it != neighbors.end(); ++it)
+            {
+              if (mg_dof.locally_owned_mg_dofs_per_processor(dofpair->level)[*it].is_element(dofpair->level_dof_index))
+                {
+                  send_data[*it].push_back(*dofpair);
+                  break;
+                }
+            }
+        }
+
+      // * send
+      std::vector<MPI_Request> requests;
+      {
+        for (std::set<unsigned int>::iterator it = neighbors.begin(); it != neighbors.end(); ++it)
+          {
+            requests.push_back(MPI_Request());
+            unsigned int dest = *it;
+            std::vector<dof_pair> &data = send_data[dest];
+            if (data.size())
+              MPI_Isend(&data[0], data.size()*sizeof(data[0]), MPI_BYTE, dest, 71, tria->get_communicator(), &*requests.rbegin());
+            else
+              MPI_Isend(NULL, 0, MPI_BYTE, dest, 71, tria->get_communicator(), &*requests.rbegin());
+          }
+      }
+
+      // * receive
+      {
+        typename std::vector<dof_pair> receive;
+        for (std::set<unsigned int>::iterator it = neighbors.begin(); it != neighbors.end(); ++it)
```

Comment:
if I see this right, you don't actually iterate over this array -- you only iterate as many times as there are elements in the array. in that case, no need to actually create an iterator, just run through an integer and call it `dummy` or something similar to indicate that you don't actually care about its value.